aiplatform/api/AIPlatform.Samples/GroundingVertexAiSearchSample.cs

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright 2024 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// [START generativeaionvertexai_gemini_grounding_with_vais]
+
+using Google.Cloud.AIPlatform.V1;
+using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+
+public class GroundingVertexAiSearchSample
+{
+    public async Task<string> GenerateTextWithVertexAiSearch(
+        string projectId = "your-project-id",
+        string location = "us-central1",
+        string publisher = "google",
+        string model = "gemini-1.0-pro-002",
+        string dataStoreLocation = "global",
+        string dataStoreId = "your-datastore-id")
+    {
+        var predictionServiceClient = new PredictionServiceClientBuilder
+        {
+            Endpoint = $"{location}-aiplatform.googleapis.com"
+        }.Build();
+
+
+        var generateContentRequest = new GenerateContentRequest
+        {
+            Model = $"projects/{projectId}/locations/{location}/publishers/{publisher}/models/{model}",
+            GenerationConfig = new GenerationConfig
+            {
+                Temperature = 0.0f
+            }
+        };
+
+        var content = new Content
+        {
+            Role = "USER"
+        };
+        content.Parts.AddRange(new List<Part>()
+        {
+            new()
+            {
+                Text = "How do I make an appointment to renew my driver's license?"
+            }
+        });
+        generateContentRequest.Contents.Add(content);
```

Comment:
I'm not sure why you're not doing it all in one statement:

```csharp
var generateContentRequest = new GenerateContentRequest
{
    Model = $"projects/{projectId}/locations/{location}/publishers/{publisher}/models/{model}",
    GenerationConfig = new GenerationConfig
    {
        Temperature = 0.0f,
    },
    Parts =
    {
        new Content
        {
            Role = "USER",
            Parts = { new Part { Text = "How do I make an appointment to renew my driver's license?" } }
        }
    },
    Tools =
    {
        new Tool
        {
            Retrieval = new Retrieval
            {
                VertexAiSearch = new VertexAISearch
                {
                    Datastore = $"projects/{projectId}/locations/{dataStoreLocation}/collections/default_collection/dataStores/{dataStoreId}"
                }
            }
        }
    }
};
```

Reply:
IIRC, Parts and Tools are read-only?

Reply:
That doesn't stop you from using them in collection initializers. See https://cloud.google.com/dotnet/docs/reference/help/protobuf

Reply:
(Basically, the compiler generates calls to fetch the property value and call Add on the collection.)

Reply:
Ah, I got bitten by this before. I do agree it looks cleaner and I'll make the change and I'll probably send another PR for all other samples to do the same. But, it still doesn't make sense to me that a read only field can be set this way. Is that a C# thing or a protobuf thing?

Reply:
It's a C# feature (and has been since C# 3). You're not *setting* the property (that would fail to compile) - you're just *initializing* the existing collection.

Reply:
I see, it's happening at construction time. Thanks, I'll make the changes

Reply:
Made the changes, thanks for the suggestion!